### ISSUE TITLE
Closures in Channel.map no longer seem to be able to include calls to custom methods

### ISSUE THREAD

**Issue by cschu** [nextflow-io/nextflow#2464]
### Expected behavior and actual behavior

In version `21.04.0.5552` (installed via bioconda), I've successfully called a custom function/method in a closure supplied to a `.map`-call during Channel creation. In nextflow version `21.10.0.5640` (also via bioconda), this is no longer possible (s. error output below). If I move the method body from the method into the closure, it works again.

### Steps to reproduce the problem

```
nextflow.enable.dsl = 2

def classify_sample(sample, files) {

    def meta = [:]
    meta.is_paired = (files instanceof Collection && files.size() == 2)
    meta.id = sample

    return [meta, files]
}

workflow {
  fastq_ch = Channel
        .fromPath(params.input_dir + "/**_*[12].{fastq,fq,fastq.gz,fq.gz}")
        .map { file ->
            def sample = file.name.replaceAll(/.(fastq|fq)(.gz)?$/, "")
            sample = sample.replaceAll(/_R?[12]$/, "")
            return tuple(sample, file)
        }
        .groupTuple(sort: true)
        .map { classify_sample(it[0], it[1]) }
        .view() 
}
```

### Program output 

On the shell: Unknown method invocation `classify_sample` on process type

In the `.nextflow.log`:

```
Nov-24 12:58:25.982 [Actor Thread 12] ERROR nextflow.extension.OperatorEx - @unknown
groovy.lang.MissingMethodException: No signature of method: nextflow.script.WorkflowDef.classify_sample() is applicable for argument types: ([Ljava.lang.Object;) values: [[[SRR5866444, [PATH-REMOVED/input/5367_3/SRR5866444_1.fastq.gz, ...]]]
]
        at nextflow.script.WorkflowParamsResolver.invokeMethod(WorkflowDef.groovy:217)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
        at org.codehaus.groovy.runtime.metaclass.TransformMetaMethod.invoke(TransformMetaMethod.java:55)
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass$1.invoke(ClosureMetaClass.java:215)
        at org.codehaus.groovy.runtime.metaclass.TransformMetaMethod.doMethodInvoke(TransformMetaMethod.java:62)
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:362)
        at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.callCurrent(PogoMetaClassSite.java:61)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:51)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:171)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:185)
        at Script_a91288fe$_runScript_closure10$_closure28$_closure30.doCall(Script_a91288fe:311)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
        at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:38)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
        at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:53)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
        at nextflow.extension.MapOp$_apply_closure1.doCall(MapOp.groovy:57)
        at jdk.internal.reflect.GeneratedMethodAccessor97.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:564)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
        at groovy.lang.Closure.call(Closure.java:412)
        at groovyx.gpars.dataflow.operator.DataflowOperatorActor.startTask(DataflowOperatorActor.java:120)
        at groovyx.gpars.dataflow.operator.DataflowOperatorActor.onMessage(DataflowOperatorActor.java:108)
        at groovyx.gpars.actor.impl.SDAClosure$1.call(SDAClosure.java:43)
        at groovyx.gpars.actor.AbstractLoopingActor.runEnhancedWithoutRepliesOnMessages(AbstractLoopingActor.java:293)
        at groovyx.gpars.actor.AbstractLoopingActor.access$400(AbstractLoopingActor.java:30)
        at groovyx.gpars.actor.AbstractLoopingActor$1.handleMessage(AbstractLoopingActor.java:93)
        at groovyx.gpars.util.AsyncMessagingCore.run(AsyncMessagingCore.java:132)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:844)
Nov-24 12:58:26.012 [Actor Thread 12] DEBUG nextflow.Session - Session aborted -- Cause: No signature of method: nextflow.script.WorkflowDef.classify_sample() is applicable for argument types: ([Ljava.lang.Object;) values: [[[SRR5866444, [PATH-REMOVED/input/5367_3/SRR5866444_1.fastq.gz, ...]]]]
```

### Environment 

* Nextflow version: 21.10.0.5640 (bioconda)
* Java version: Groovy 3.0.9 on OpenJDK 64-Bit Server VM 10.0.2+13
* Operating system: Linux 3.10.0-1160.36.2.el7.x86_64
* Bash version: GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)

#### Working Environment

* Nextflow version: 21.04.0 build 5552 (bioconda)
* Java version: Groovy 3.0.7 on OpenJDK 64-Bit Server VM 11.0.9.1-internal+0-adhoc..src
* Operating system: Linux 3.10.0-1160.36.2.el7.x86_64
* Bash version: GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)

**Comment by dennishendriksen** [comment]
I'm running in a similar issue after upgrading Nextflow 21.04.3 to 21.10.3:

```
groovy.lang.MissingMethodException: No signature of method: nextflow.script.WorkflowDef.nr_records() is applicable for argument types: ([Ljava.lang.Object;) values: [[/gro
ups/solve-rd/tmp10/projects/vip/nextflow/git/vip/test/output/test_snv/.nxf_work/92/1158ff50d7175029ff592f11a241f3/snv_prepared.vcf.gz.stats]]
        at nextflow.script.WorkflowParamsResolver.invokeMethod(WorkflowDef.groovy:217)
```

```
def nr_records(statsFilePath) {
  ...
}
```

```
 channel.fromPath(params.input) \
      | prepare \
      | branch {
          small: nr_records(it.last()) <= params.chunk_size
          large: true
        }
      ...
```

```
process prepare {
  ...
  output:
    tuple ..., path("${vcfOutputPath}.stats")
  ...
```

I've tried updating in the caller and callee in various ways to work around this issue without success.

**Comment by pditommaso** [comment]
This looks the same as #2469

**Comment by cschu** [comment]
> This looks the same as #2469

Yes, it appears so. I can also confirm that my 'broken' example above actually works (i.e. if the method is defined within the same file), so it seems to be the include.

**Comment by pditommaso** [comment]
Solved by #2469. Available as of 21.10.4  https://github.com/nextflow-io/nextflow/releases/tag/v21.10.4
